### PR TITLE
aflplusplus: 2.65c -> 2.66c

### PR DESCRIPTION
**Pull request by Mindavi**
Added cmocka back in and add a new dependency, python-setuptools.
Cmocka is added back because the unit tests would otherwise fail.

###### Motivation for this change

Update of the package upstream

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
  - [x] afl-clang-fast
  - [x] afl-clang-fast++
  - [x] afl-fuzz
  - [x] afl-system-config
  - [ ] afl-analyze
  - [ ] afl-cgroup
  - [ ] afl-cmin
  - [ ] afl-cmin.bash
  - [ ] afl-g++
  - [ ] afl-gcc
  - [ ] afl-gcc-fast
  - [ ] afl-g++-fast
  - [x] afl-gotcpu
  - [ ] afl-plot
  - [ ] afl-qemu-trace
  - [ ] afl-showmap
  - [ ] afl-tmin
  - [ ] afl-whatup  
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I have only tested this lightly, so leaving it as a draft for now. Please let me know what your experience is if you're reading this and have tested anything :+1:.

cc @risicle

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -617,6 +617,7 @@ in
     clang = clang_9;
     llvm = llvm_9;
     python = python37;
+    pythonPackages = python37Packages;
```
Comment (risicle) [inline pkgs/top-level/all-packages.nix:620]: You can actually access a `python`'s packages through `python.pkgs`, which would be more elegant in this case than adding a new arg.

**Comment by risicle** [comment]
Hmm one of the `cmocka` unit tests fails for me:

```
[*] Execution cmocka Unit-Tests 
make: Entering directory '/build/source'
./test/unittests/unit_maybe_alloc
[==========] Running 5 test(s).
[ RUN      ] test_null_allocs
[       OK ] test_null_allocs
[ RUN      ] test_nonpow2_size
[       OK ] test_nonpow2_size
[ RUN      ] test_zero_size
)B
[-] PROGRAM ABORT : cannot grow to non-positive size
         Location : ck_maybe_grow(), include/alloc-inl.h:830
```

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (Mindavi) [reply pkgs/top-level/all-packages.nix:620]: That's nice, I wasn't aware of that :)

**Comment by Mindavi** [comment]
IIf I look at the output it says it should actually abort like that, so I guess it's okay that that output is shown.

**Comment by risicle** [comment]
No, but it does actually fail for me.

**Comment by Mindavi** [comment]
So my output is this:

```
[*] Execution cmocka Unit-Tests 
make: Entering directory '/build/source'
./test/unittests/unit_maybe_alloc
[==========] Running 5 test(s).
[ RUN      ] test_null_allocs
[       OK ] test_null_allocs
[ RUN      ] test_nonpow2_size
[       OK ] test_nonpow2_size
[ RUN      ] test_zero_size
)B
[-] PROGRAM ABORT : cannot grow to non-positive size
         Location : ck_maybe_grow(), include/alloc-inl.h:830

Expected assertion 0 occurred
[       OK ] test_zero_size
[ RUN      ] test_unchanged_size
[       OK ] test_unchanged_size
[ RUN      ] test_grow_multiple
[       OK ] test_grow_multiple
[==========] 5 test(s) run.
[  PASSED  ] 5 test(s).
./test/unittests/unit_preallocable
[==========] Running 2 test(s).
[ RUN      ] test_alloc_free
[       OK ] test_alloc_free
[ RUN      ] test_prealloc_overflow
[       OK ] test_prealloc_overflow
[==========] 2 test(s) run.
[  PASSED  ] 2 test(s).
./test/unittests/unit_list
[==========] Running 3 test(s).
[ RUN      ] test_contains
[       OK ] test_contains
[ RUN      ] test_foreach
[       OK ] test_foreach
[ RUN      ] test_long_list
[       OK ] test_long_list
[==========] 3 test(s) run.
[  PASSED  ] 3 test(s).
./test/unittests/unit_rand
[==========] Running 2 test(s).
[ RUN      ] test_rand_0
[       OK ] test_rand_0
[ RUN      ] test_rand_below
[       OK ] test_rand_below
[==========] 2 test(s) run.
[  PASSED  ] 2 test(s).
./test/unittests/unit_hash
[==========] Running 1 test(s).
[ RUN      ] test_hash
[       OK ] test_hash
[==========] 1 test(s) run.
[  PASSED  ] 1 test(s).
make: Leaving directory '/build/source'
[*] all test cases completed.
[-] not all test cases were executed
[+] all tests were successful :-)
```

Does yours say that tests have failed at the end?

Note the **Expected assertion 0 occurred** after the 'PROGRAM ABORT'.

**Comment by Mindavi** [comment]
If the cmocka tests are not working, I might remove them. I haven't seen the issue though, so I'd like feedback on what is going wrong at your side. The 'PROGRAM ABORT' is shown for me as well, but it does not fail the tests (and seems to be correct anyhow).

**Comment by risicle** [comment]
This is working for me now.

**Comment by risicle** [comment]
Could you squash down your commits and rebase?

I should probably check it still works with wine and i686, which means a massive download for me...

**Comment by Mindavi** [comment]
If you could explain how to do that, I'd be happy to run some checks. Or is testing by yourself easier?

I'll squash the commits down and rebase so it's one clean commit when I find the time :+1:

**Comment by risicle** [comment]
Hmmmmmm... tests are failing for me on i686:

```
[*] Testing the CC wrapper and instrumentation output...
unset AFL_USE_ASAN AFL_USE_MSAN AFL_INST_RATIO; AFL_QUIET=1 AFL_PATH=. AFL_LLVM_LAF_SPLIT_SWITCHES=1 AFL_LLVM_LAF_TRANSFORM_COMPARES=1 AFL_LLVM_LAF_SPLIT_COMPARES=1 ../afl-clang-fast -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -Wno-pointer-sign -I ../include/ -DAFL_PATH=\"/nix/store/gi82s5h35sly0y0am2c9w5c5ng40yghx-aflplusplus-2.66c/lib/afl\" -DBIN_PATH=\"/nix/store/gi82s5h35sly0y0am2c9w5c5ng40yghx-aflplusplus-2.66c/bin\" -DLLVM_BINDIR=\"/nix/store/p347p4hhighw33zx8bhmvybwl2af0c39-llvm-9.0.1/bin\" -DVERSION=\"++2.66c\" -DLLVM_LIBDIR=\"/nix/store/l30kxpx9b5176mmj5nly8lzzzk1zralb-llvm-9.0.1-lib/lib\" -DLLVM_VERSION=\"9.0.1\" -DAFL_CLANG_FLTO=\"-flto=full\" -DAFL_REAL_LD=\"\" -DAFL_CLANG_FUSELD=\"1\" -DCLANG_BIN=\"clang\" -DCLANGPP_BIN=\"clang++\" -DUSE_BINDIR=0 -Wno-unused-function ../test-instr.c -o test-instr
/nix/store/j7la2zygdbwg2yz7x20g28gyp9apni42-binutils-2.31.1/bin/ld: cannot find /nix/store/n6455n8zivqs6gfydxif4qi83smslqwi-clang-wrapper-9.0.1/resource-root/lib/linux/libclang_rt.ubsan_standalone-i386.a: No such file or directory
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
```

The i686 package is available under `pkgsi686Linux.aflplusplus` if you want to see for yourself.

**Comment by risicle** [comment]
I wonder if we just don't build ubsan for i686 clang?

**Comment by risicle** [comment]
I investigated this further and that does seem to be the case: #97688

Wonder if we should try disabling the ubsan/asan tests for i686.

**Comment by Mindavi** [comment]
Yeah maybe that would be best. It seems like ubsan is used by default now in aflplusplus (at least, I've seen ubsan warnings without explicitly enabling it) builds, so I think we should disable the auto-enabling of it.

**Comment by Mindavi** [comment]
Mm, after the rebase with master I get lots of test failures too now, that's no good.

**Comment by Mindavi** [comment]
Gonna close this, I guess it would be good to try 3.00c in the future. As packaging (at least with most/all features) is a huge effort I'm afraid that might take a while.

I think the upstream system of creating makefiles that try to be autoconf or cmake is not the best approach, it might be good to have a healthy discussion there to see if they can approach this differently. Currently we have to patch so much and work around all the idiosyncrasies in the build system that it's really a hassle to get it properly working. There's also a lot of stuff to test to ensure it's all ok.

Automated tests in nixos might help packaging as well, but that requires some setup as well and I'm not too well versed in the testing system. However, it'll make figuring out updates somewhat smoother as we can test some different setups there.

**Pull request closed without merging**